tfx/orchestration/experimental/kubernetes/kubernetes_remote_runner_test.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Kubernetes TFX runner for out-of-cluster orchestration."""
+
+import tensorflow as tf
+from ml_metadata.proto import metadata_store_pb2
+from google.protobuf import json_format
+import json
+
+from tfx import types
+from tfx.components.base import base_component
+from tfx.components.base import base_executor
+from tfx.components.base import executor_spec
+from tfx.orchestration import pipeline
+from tfx.orchestration.experimental.kubernetes import kubernetes_remote_runner
+from tfx.types.component_spec import ChannelParameter
+from tfx.utils import json_utils
+
+
+class _ArtifactTypeA(types.Artifact):
+  TYPE_NAME = 'ArtifactTypeA'
+
+
+class _ArtifactTypeB(types.Artifact):
+  TYPE_NAME = 'ArtifactTypeB'
+
+
+class _FakeComponentSpecA(types.ComponentSpec):
+  PARAMETERS = {}
+  INPUTS = {}
+  OUTPUTS = {'output': ChannelParameter(type=_ArtifactTypeA)}
+
+
+class _FakeComponentSpecB(types.ComponentSpec):
+  PARAMETERS = {}
+  INPUTS = {'a': ChannelParameter(type=_ArtifactTypeA)}
+  OUTPUTS = {'output': ChannelParameter(type=_ArtifactTypeB)}
+
+class _FakeComponent(base_component.BaseComponent):
+  SPEC_CLASS = types.ComponentSpec
+  EXECUTOR_SPEC = executor_spec.ExecutorClassSpec(base_executor.BaseExecutor)
+
+  def __init__(self, spec: types.ComponentSpec):
+    instance_name = spec.__class__.__name__.replace(
+        '_FakeComponentSpec', '').lower()
+    super(_FakeComponent, self).__init__(spec=spec, instance_name=instance_name)
+
+class KubernetesRemoteRunnerTest(tf.test.TestCase):
+
+  def setUp(self):
+    self.component_a = _FakeComponent(
+        _FakeComponentSpecA(output=types.Channel(type=_ArtifactTypeA)))
+    self.component_b = _FakeComponent(
+        _FakeComponentSpecB(
+            a=self.component_a.outputs['output'],
+            output=types.Channel(type=_ArtifactTypeB)))
```

Comment:
Can you test some more complicated pipeline topologies?  Right now, you only test with two nodes.  Maybe add another component that takes as inputs A.output and B.output.

Reply:
Added an extra node. Also added a test for deserialization.